rllib/agents/trainer.py

```diff
@@ -764,6 +764,74 @@ def env_creator_from_classpath(env_context):
                 "`callbacks` must be a callable method that "
                 "returns a subclass of DefaultCallbacks, got {}".format(
                     self.config["callbacks"]))
+
+        # Update the user provided callbacks with RE3 specific changes
+        if self.config["exploration_config"].get("type", None) == "RE3":
+            from ray.rllib.evaluation.postprocessing import Postprocessing
+            from ray.rllib.utils.exploration.random_encoder import (
+                update_beta,
+                compute_states_entropy,
+                MovingMeanStd,
+            )
+            from ray.rllib.policy.sample_batch import SampleBatch
+
+            embeds_dim = self.config["exploration_config"].get(
+                "embeds_dim", 128)
+            k_nn = self.config["exploration_config"].get("k_nn", 50)
+            beta = self.config["exploration_config"].get("beta", 0.1)
+            rho = self.config["exploration_config"].get("rho", 0.0001)
+            beta_schedule = self.config["exploration_config"].get(
+                "beta_schedule", "constant")
+
+            class UpdateCallbacks(self.config["callbacks"]):
+                """Update input callbacks method to mutate batch with states entropy rewards."""
+
+                _step = 0
+
+                def __init__(self, *args, **kwargs):
+                    self._rms = MovingMeanStd()
+                    super().__init__(*args, **kwargs)
+
+                def on_learn_on_batch(
+                        self,
+                        *,
+                        policy: Policy,
+                        train_batch: SampleBatch,
+                        result: dict,
+                        **kwargs,
+                ):
+                    states_entropy = compute_states_entropy(
+                        train_batch[SampleBatch.OBS_EMBEDS], embeds_dim, k_nn)
```

Comment:
this feels a bit weird.
I guess to calculate the knn distance, you want to use a batch that's randomly sampled from the ReplayBuffer? just so you can measure how different current sample batch is from all the things you have been seeing (RB)?
as written, this knn thing is calculated from a single batch. so naturally, you are gonna get a lot of similar samples from consecutive steps?

Reply:
Yes, KNN is computed for a single batch and we are assuming that RLLib has sampled that batch randomly from replay buffer.

Reply:
KNN distance and the randomness is limited to single batch only, not the full replay buffer.